File: src/utils/earTrainingPrecisionLoop.ts
import { PRECISION_NOTE_FALL_LEAD_SEC } from '@/components/piano/PrecisionNotesRenderer';
import type { PrecisionNote } from '@/utils/earTrainingPrecisionNotes';

export type LoopTransposeDirection = 'down' | 'up' | 'fourthUp' | 'none';

const LOOP_TRANSPOSE_STEP_SEMITONES: Record<LoopTransposeDirection, number> = {
  none: 0,
  down: -1,
  up: 1,
  fourthUp: 5,
};

export interface PrecisionLoopWindow {
  startMeasure: number;
  endMeasure: number;
  startSec: number;
  endSec: number;
  durationSec: number;
}

export interface PrecisionLoopPosition {
  cycleIndex: number;
  localSec: number;
}

export interface BuildLoopWindowNotesParams {
  notesBySemitone: ReadonlyMap<number, readonly PrecisionNote[]>;
  cycleIndex: number;
  windowSize: number;
  loopWindow: PrecisionLoopWindow;
  direction: LoopTransposeDirection;
  baseSemitone?: number;
  resolveCalibratedStartSec?: (startSec: number) => number;
}

export const resolveLoopWindow = (params: {
  startMeasure: number;
  endMeasure: number;
  measureDurationSec: number;
}): PrecisionLoopWindow => {
  const startMeasure = Math.max(1, Math.trunc(params.startMeasure));
  const endMeasure = Math.max(startMeasure, Math.trunc(params.endMeasure));
  const measureDurationSec = Math.max(1e-6, params.measureDurationSec);
  const startSec = (startMeasure - 1) * measureDurationSec;
  const endSec = endMeasure * measureDurationSec;
  return {
    startMeasure,
    endMeasure,
    startSec,
    endSec,
    durationSec: Math.max(1e-6, endSec - startSec),
  };
};

export const loopSemitoneForCycle = (
  cycleIndex: number,
  direction: LoopTransposeDirection,
  baseSemitone = 0,
): number => {
  const safeCycle = Math.max(0, Math.trunc(cycleIndex));
  const step = LOOP_TRANSPOSE_STEP_SEMITONES[direction] * safeCycle;
  const raw = baseSemitone + step;
  let wrapped = ((raw % 12) + 12) % 12;
  if (wrapped > 6 || (wrapped === 6 && raw < 0)) {
    wrapped -= 12;
  }
  return wrapped;
};

export const loopPracticeUniqueSemitones = (
  direction: LoopTransposeDirection,
  baseSemitone = 0,
): readonly number[] => {
  if (direction === 'none') {
    return [loopSemitoneForCycle(0, direction, baseSemitone)];
  }
  const seen = new Set<number>();
  const result: number[] = [];
  for (let cycle = 0; cycle < 12; cycle += 1) {
    const semitone = loopSemitoneForCycle(cycle, direction, baseSemitone);
    if (seen.has(semitone)) {
      continue;
    }
    seen.add(semitone);
    result.push(semitone);
  }
  return result;
};

export const globalToLoopPosition = (
  globalSec: number,
  durationSec: number,
): PrecisionLoopPosition => {
  const safeDuration = Math.max(1e-6, durationSec);
  const safeGlobal = Math.max(0, globalSec);
  const cycleIndex = Math.floor(safeGlobal / safeDuration);
  const localSec = safeGlobal - cycleIndex * safeDuration;
  return { cycleIndex, localSec };
};

export const loopCycleWindowSize = (durationSec: number): number => {
  const safeDuration = Math.max(1e-6, durationSec);
  return Math.max(2, Math.ceil(PRECISION_NOTE_FALL_LEAD_SEC / safeDuration) + 1);
};

const noteOverlapsLoopWindow = (
  note: PrecisionNote,
  loopWindow: PrecisionLoopWindow,
): boolean => {
  const noteEndSec = note.startSec + note.durationSec;
  return noteEndSec > loopWindow.startSec + 1e-9 && note.startSec < loopWindow.endSec - 1e-9;
};

export const buildLoopWindowNotes = ({
  notesBySemitone,
  cycleIndex,
  windowSize,
  loopWindow,
  direction,
  baseSemitone = 0,
  resolveCalibratedStartSec = (value) => value,
}: BuildLoopWindowNotesParams): PrecisionNote[] => {
  const safeCycle = Math.max(0, Math.trunc(cycleIndex));
  const safeWindowSize = Math.max(1, Math.trunc(windowSize));
  const built: PrecisionNote[] = [];

  for (let offset = 0; offset < safeWindowSize; offset += 1) {
    const cycle = safeCycle + offset;
    const semitone = loopSemitoneForCycle(cycle, direction, baseSemitone);
    const baseNotes = notesBySemitone.get(semitone) ?? [];
    const globalCycleOffsetSec = cycle * loopWindow.durationSec;

    for (const note of baseNotes) {
      if (!noteOverlapsLoopWindow(note, loopWindow)) {
        continue;
      }
      const localStartSec = note.startSec - loopWindow.startSec;
      const globalStartSec = resolveCalibratedStartSec(localStartSec + globalCycleOffsetSec);
      built.push({
        ...note,
        id: `${note.id}#c${cycle}`,
        startSec: globalStartSec,
        measureNumber: note.measureNumber,
      });
    }
  }

  built.sort((left, right) => {
    if (left.startSec !== right.startSec) {
      return left.startSec - right.startSec;
    }
    return left.id.localeCompare(right.id);
  });
  return built;
};

export const loopActiveMeasureNumber = (
  localSec: number,
  measureDurationSec: number,
  loopWindow: PrecisionLoopWindow,
  maxMeasure: number,
): number => {
  const measureInWindow = Math.floor(Math.max(0, localSec) / Math.max(1e-6, measureDurationSec)) + 1;
  const spanMeasures = loopWindow.endMeasure - loopWindow.startMeasure + 1;
  const clampedInWindow = Math.max(1, Math.min(spanMeasures, measureInWindow));
  const absoluteMeasure = loopWindow.startMeasure + clampedInWindow - 1;
  return Math.max(1, Math.min(maxMeasure, absoluteMeasure));
};

export const loopOsmdTimelineSec = (
  localSec: number,
  measureDurationSec: number,
  loopWindow: PrecisionLoopWindow,
): { measureNumber: number; phraseTimelineSec: number } => {
  const measureNumber = loopActiveMeasureNumber(
    localSec,
    measureDurationSec,
    loopWindow,
    loopWindow.endMeasure,
  );
  const measureInWindow = measureNumber - loopWindow.startMeasure + 1;
  const offsetInMeasure = localSec - (measureInWindow - 1) * measureDurationSec;
  const phraseTimelineSec = (measureNumber - 1) * measureDurationSec + offsetInMeasure;
  return { measureNumber, phraseTimelineSec };
};

File: ios/Jazzify/EarTraining/Engine/EarTrainingPrecisionLoop.swift
import Foundation

enum EarTrainingLoopTransposeDirection: String, CaseIterable, Sendable {
    case down
    case up
    case fourthUp
    case none
}

struct EarTrainingPrecisionLoopWindow: Equatable, Sendable {
    let startMeasure: Int
    let endMeasure: Int
    let startSec: Double
    let endSec: Double
    let durationSec: Double
}

struct EarTrainingPrecisionLoopPosition: Equatable, Sendable {
    let cycleIndex: Int
    let localSec: Double
}

struct EarTrainingPrecisionLoopOsmdTimeline: Equatable, Sendable {
    let measureNumber: Int
    let phraseTimelineSec: Double
}

enum EarTrainingPrecisionLoop {
    static let loopBaseSemitoneMin = -6
    static let loopBaseSemitoneMax = 5

    static func resolveLoopWindow(
        startMeasure: Int,
        endMeasure: Int,
        measureDurationSec: Double
    ) -> EarTrainingPrecisionLoopWindow {
        let safeStart = max(1, startMeasure)
        let safeEnd = max(safeStart, endMeasure)
        let safeMeasureDuration = max(1e-6, measureDurationSec)
        let startSec = Double(safeStart - 1) * safeMeasureDuration
        let endSec = Double(safeEnd) * safeMeasureDuration
        return EarTrainingPrecisionLoopWindow(
            startMeasure: safeStart,
            endMeasure: safeEnd,
            startSec: startSec,
            endSec: endSec,
            durationSec: max(1e-6, endSec - startSec)
        )
    }

    static func loopSemitoneForCycle(
        _ cycleIndex: Int,
        direction: EarTrainingLoopTransposeDirection,
        baseSemitone: Int = 0
    ) -> Int {
        let safeCycle = max(0, cycleIndex)
        let step: Int
        switch direction {
        case .none:
            step = 0
        case .down:
            step = -safeCycle
        case .up:
            step = safeCycle
        case .fourthUp:
            step = safeCycle * 5
        }
        let raw = baseSemitone + step
        var wrapped = ((raw % 12) + 12) % 12
        if wrapped > 6 || (wrapped == 6 && raw < 0) {
            wrapped -= 12
        }
        return wrapped
    }

    static func loopPracticeUniqueSemitones(
        direction: EarTrainingLoopTransposeDirection,
        baseSemitone: Int = 0
    ) -> [Int] {
        if direction == .none {
            return [loopSemitoneForCycle(0, direction: direction, baseSemitone: baseSemitone)]
        }
        var seen = Set<Int>()
        var result: [Int] = []
        for cycle in 0..<12 {
            let semitone = loopSemitoneForCycle(cycle, direction: direction, baseSemitone: baseSemitone)
            if seen.contains(semitone) {
                continue
            }
            seen.insert(semitone)
            result.append(semitone)
        }
        return result
    }

    static func globalToLoopPosition(
        _ globalSec: Double,
        durationSec: Double
    ) -> EarTrainingPrecisionLoopPosition {
        let safeDuration = max(1e-6, durationSec)
        let safeGlobal = max(0, globalSec)
        let cycleIndex = Int(floor(safeGlobal / safeDuration))
        let localSec = safeGlobal - Double(cycleIndex) * safeDuration
        return EarTrainingPrecisionLoopPosition(cycleIndex: cycleIndex, localSec: localSec)
    }

    static func loopCycleWindowSize(durationSec: Double) -> Int {
        let safeDuration = max(1e-6, durationSec)
        return max(2, Int(ceil(EarTrainingPrecisionNotes.fallLeadSec / safeDuration)) + 1)
    }

    static func buildLoopWindowNotes(
        notesBySemitone: [Int: [EarTrainingPrecisionNote]],
        cycleIndex: Int,
        windowSize: Int,
        loopWindow: EarTrainingPrecisionLoopWindow,
        direction: EarTrainingLoopTransposeDirection,
        baseSemitone: Int = 0,
        resolveCalibratedStartSec: (Double) -> Double = { $0 }
    ) -> [EarTrainingPrecisionNote] {
        let safeCycle = max(0, cycleIndex)
        let safeWindowSize = max(1, windowSize)
        var built: [EarTrainingPrecisionNote] = []

        for offset in 0..<safeWindowSize {
            let cycle = safeCycle + offset
            let semitone = loopSemitoneForCycle(cycle, direction: direction, baseSemitone: baseSemitone)
            let baseNotes = notesBySemitone[semitone] ?? []
            let globalCycleOffsetSec = Double(cycle) * loopWindow.durationSec

            for note in baseNotes {
                guard noteOverlapsLoopWindow(note, loopWindow: loopWindow) else { continue }
                let localStartSec = note.startSec - loopWindow.startSec
                let globalStartSec = resolveCalibratedStartSec(localStartSec + globalCycleOffsetSec)
                built.append(
                    EarTrainingPrecisionNote(
                        id: "\(note.id)#c\(cycle)",
                        midi: note.midi,
                        startSec: globalStartSec,
                        durationSec: note.durationSec,
                        isBlackKey: note.isBlackKey,
                        measureNumber: note.measureNumber,
                        isShortNote: note.isShortNote
                    )
                )
            }
        }

        built.sort { left, right in
            if left.startSec != right.startSec {
                return left.startSec < right.startSec
            }
            return left.id < right.id
        }
        return built
    }

    static func loopActiveMeasureNumber(
        localSec: Double,
        measureDurationSec: Double,
        loopWindow: EarTrainingPrecisionLoopWindow,
        maxMeasure: Int
    ) -> Int {
        let safeMeasureDuration = max(1e-6, measureDurationSec)
        let measureInWindow = Int(floor(max(0, localSec) / safeMeasureDuration)) + 1
        let spanMeasures = loopWindow.endMeasure - loopWindow.startMeasure + 1
        let clampedInWindow = max(1, min(spanMeasures, measureInWindow))
        let absoluteMeasure = loopWindow.startMeasure + clampedInWindow - 1
        return max(1, min(maxMeasure, absoluteMeasure))
    }

    static func loopOsmdTimelineSec(
        localSec: Double,
        measureDurationSec: Double,
        loopWindow: EarTrainingPrecisionLoopWindow
    ) -> EarTrainingPrecisionLoopOsmdTimeline {
        let measureNumber = loopActiveMeasureNumber(
            localSec: localSec,
            measureDurationSec: measureDurationSec,
            loopWindow: loopWindow,
            maxMeasure: loopWindow.endMeasure
        )
        let measureInWindow = measureNumber - loopWindow.startMeasure + 1
        let offsetInMeasure = localSec - Double(measureInWindow - 1) * measureDurationSec
        let phraseTimelineSec = Double(measureNumber - 1) * measureDurationSec + offsetInMeasure
        return EarTrainingPrecisionLoopOsmdTimeline(
            measureNumber: measureNumber,
            phraseTimelineSec: phraseTimelineSec
        )
    }

    static func buildPrecisionNotesBySemitone(
        xmlText: String?,
        midiData: Data?,
        bpm: Int,
        beatsPerMeasure: Int,
        isSwing: Bool,
        direction: EarTrainingLoopTransposeDirection,
        baseSemitone: Int = 0,
        resolveCalibratedStartSec: @escaping (Double) -> Double,
        practiceMode: Bool,
        practiceSpeedPercent: Int,
        classificationBpm: Int
    ) -> [Int: [EarTrainingPrecisionNote]] {
        let semitones = loopPracticeUniqueSemitones(direction: direction, baseSemitone: baseSemitone)
        var result: [Int: [EarTrainingPrecisionNote]] = [:]

        for semitone in semitones {
            var builtNotes: [EarTrainingPrecisionNote] = []
            if let midiData {
                if let buildResult = try? EarTrainingPrecisionMidi.buildFromMidi(
                    data: midiData,
                    bpm: bpm,
                    transposeOffset: semitone
                ) {
                    builtNotes = buildResult.notes
                }
            } else if let xmlText {
                builtNotes = EarTrainingPrecisionNotes.buildFromMusicXml(
                    musicXmlText: xmlText,
                    bpm: bpm,
                    beatsPerMeasure: beatsPerMeasure,
                    transposeOffset: semitone,
                    isSwing: isSwing
                ).notes
            }
            let calibrated = EarTrainingPrecisionNotes.calibrateNotes(
                notes: builtNotes,
                resolveCalibratedStartSec: resolveCalibratedStartSec,
                practiceMode: practiceMode,
                practiceSpeedPercent: practiceSpeedPercent,
                classificationBpm: classificationBpm
            )
            result[semitone] = calibrated
        }
        return result
    }

    private static func noteOverlapsLoopWindow(
        _ note: EarTrainingPrecisionNote,
        loopWindow: EarTrainingPrecisionLoopWindow
    ) -> Bool {
        let noteEndSec = note.startSec + note.durationSec
        return noteEndSec > loopWindow.startSec + 1e-9 && note.startSec < loopWindow.endSec - 1e-9
    }
}
